File: service/src/test/java/hu/bsstudio/robonaut/race/speed/timer/BroadcastingSpeedTimerServiceTest.java
package hu.bsstudio.robonaut.race.speed.timer;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.mockito.MockitoAnnotations.openMocks;

import hu.bsstudio.robonaut.race.speed.timer.model.SpeedTimer;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

final class BroadcastingSpeedTimerServiceTest {

    private static final SpeedTimer SPEED_TIMER = new SpeedTimer(0, null);
    private static final String ROUTING_KEY = "speed.timer";

    private BroadcastingSpeedTimerService underTest;

    @Mock
    private RabbitTemplate mockTemplate;
    @Mock
    private SpeedTimerService mockService;

    @BeforeEach
    void setUp() {
        openMocks(this);
        this.underTest = new BroadcastingSpeedTimerService(mockTemplate, mockService);
    }

    @Test
    void shouldReturnSpeedTimerFromUnderLyingServiceAndSendItWhenTimerIsUpdated() {
        when(mockService.updateTimer(SPEED_TIMER))
            .thenReturn(Mono.just(SPEED_TIMER));

        Mono.just(SPEED_TIMER)
            .flatMap(underTest::updateTimer)
            .as(StepVerifier::create)
            .expectNext(SPEED_TIMER)
            .verifyComplete();
        verify(mockTemplate).convertAndSend(ROUTING_KEY, SPEED_TIMER);
    }
}

File: service/src/test/kotlin/hu/bsstudio/robonaut/race/speed/timer/BroadcastingSpeedTimerServiceTest.kt
package hu.bsstudio.robonaut.race.speed.timer

import hu.bsstudio.robonaut.common.model.TimerAction
import hu.bsstudio.robonaut.race.speed.timer.model.SpeedTimer
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.amqp.rabbit.core.RabbitTemplate
import reactor.core.publisher.Mono
import reactor.test.StepVerifier

internal class BroadcastingSpeedTimerServiceTest {

    @MockK
    private lateinit var mockTemplate: RabbitTemplate

    @MockK
    private lateinit var mockService: SpeedTimerService

    private lateinit var underTest: BroadcastingSpeedTimerService

    @BeforeEach
    internal fun setUp() {
        MockKAnnotations.init(this)
        underTest = BroadcastingSpeedTimerService(mockTemplate, mockService)
    }

    @Test
    internal fun shouldReturnSpeedTimerFromUnderLyingServiceAndSendItWhenTimerIsUpdated() {
        every { mockService.updateTimer(SPEED_TIMER) } returns Mono.just(SPEED_TIMER)
        every { mockTemplate.convertAndSend(ROUTING_KEY, SPEED_TIMER) } returns Unit

        Mono.just(SPEED_TIMER)
            .flatMap(underTest::updateTimer)
            .let(StepVerifier::create)
            .expectNext(SPEED_TIMER)
            .verifyComplete()
    }

    companion object {
        private val SPEED_TIMER = SpeedTimer(0, TimerAction.START)
        private const val ROUTING_KEY = "speed.timer"
    }
}
